tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/passes/TemporalPassImpl.scala

```diff
@@ -0,0 +1,149 @@
+
+    Right(newModule)
+  }
+
+  def temporalToInvariants(
+      module: TlaModule,
+      temporalProperties: List[String],
+      init: String,
+      next: String): TlaModule = {
+    val levelFinder = new TlaLevelFinder(module)
+
+    val temporalFormulas = temporalProperties
+      .map(propName => {
+        module.declarations.find(_.name == propName)
+      })
+      .filter(propOption =>
+        propOption match {
+          case Some(inv: TlaOperDecl) if inv.formalParams.isEmpty =>
+            // either a state invariant, or an action invariant
+            val level = levelFinder.getLevelOfDecl(inv)
+            level == TlaLevelTemporal
+          case _ => false
+        })
+      .map(propOption => propOption.get.asInstanceOf[TlaOperDecl])
+
+    if (temporalFormulas.isEmpty) {
+      logger.info("  > No temporal properties found, nothing to encode")
+      module
+    } else {
+      logger.info(s"  > Found ${temporalFormulas.length} temporal properties")
+      logger.info(s"  > Adding logic for loop finding")
+
+      val initDecl = module.declarations.find(_.name == init) match {
+        case Some(initDecl: TlaOperDecl) =>
+          val nparams = initDecl.formalParams.length
+          if (nparams != 0) {
+            val message =
+              s"Expected init predicate $init to have 0 params, found $nparams parameters"
+            throw new TlaInputError(message, Some(initDecl.body.ID))
+          }
+          initDecl
+        case None =>
+          val message = (s"init predicate named `${init}` not found")
+          throw new TlaInputError(message)
+        case _ =>
+          val message = (s"Expected to find a predicate named `${init}` but did not")
+          throw new TlaInputError(message)
+      }
+
+      val nextDecl = module.declarations.find(_.name == next) match {
+        case Some(nextDecl: TlaOperDecl) =>
+          val nparams = nextDecl.formalParams.length
+          if (nparams != 0) {
+            val message =
+              s"Expected next predicate $next to have 0 params, found $nparams parameters"
+            throw new TlaInputError(message, Some(nextDecl.body.ID))
+          }
+          nextDecl
+        case None =>
+          val message = (s"next predicate named `${next}` not found")
+          throw new TlaInputError(message)
+        case _ =>
+          val message = (s"Expected to find a predicate named `${next}` but did not")
+          throw new TlaInputError(message)
+      }
+
+      val loopEncoder = new LoopEncoder(tracker)
+
+      val loopModWithPreds =
+        loopEncoder.addLoopLogic(module, initDecl, nextDecl)
+
+      // let-in expressions are hard to handle in temporal formulas,
+      // so they are inlined here
+      val inliner = new Inliner(tracker, renaming, keepNullary = false)
+      val scope = Map[String, TlaOperDecl](temporalFormulas.map(operDecl => (operDecl.name, operDecl)): _*)
+      val transformation = inliner.transform(scope)
```

Comment:
I assume there's a proper inliner pass before this (the inliner here is merely because
I don't want to keep nullary in temporal formulas, but they are fine outside of this),
so I think such dependencies will already be gone.